### DISCUSSION TITLE
Documenting Wagtail's internals

### DISCUSSION THREAD

**Discussion by laymonage** [wagtail/wagtail#12546] [Development]
## Gap between our current practices vs the documentation

The recent releases of Wagtail introduced new features and UI enhancements across the admin. Whenever possible, these features were implemented with reusability and extensibility in mind. 

This seems to work out quite well: as evidenced in our rollout of "Universal Listings" (#10446), we were able to reuse much of the code across different areas of the admin (images, documents, snippets, and others). While the internal APIs to build these features are not considered stable yet, there are some "key" architectural patterns that have seen little to no changes.

For example, even though we still iterate on the UI framework for building tables (`wagtail.admin.ui.tables`), the pattern of defining columns with `Column` subclasses and passing the instances to the `Table` instance has remained the same since its introduction (https://github.com/wagtail/wagtail/commit/01d0926d956d61cc28cc33ddd89548abe2c53d74 v2.15, 3 years ago).

Despite the tables framework itself being undocumented, we inevitably mention some of its internals when it comes to customising the admin. An example of this is a mention of the `Column` class in [`ModelViewSet.list_display`](https://docs.wagtail.org/en/stable/reference/viewsets.html#wagtail.admin.viewsets.model.ModelViewSet.list_display). Aside from quick mentions like this, there's no documentation on how to use the `Column` class, what built-in subclasses are available, or how to create your own subclass. This is contrary to the fact that when the question of "how do I add a custom column that displays X?" comes up, our preferred answer is something to do with the tables UI framework (instead of, say, overriding the templates).

It hurts to say "you can achieve this cool stuff by doing ABC" but not be able to point to a relevant section for ABC in the docs, and I'm sure it's not satisfactory for developers to hear that answer 🥲 

## Why?

One of the main reasons why we haven't documented these internals is that once it's part of the public API, any breaking changes must go through the deprecation period.

With the release of Wagtail 3.0 more than two years ago, we switched to adapt SemVer more closely: deprecations will go through at least two feature releases (and will raise a warning), to be removed in the next major release (exact version TBD). This means if we start documenting the internals, further iterations will slow down as we'll need to make sure not to break things outright, and keep old code during the deprecation period (as well as documenting every change that affects the documented API).

It seems our deprecation policy hasn't been updated since then (#12543), though. It still says that deprecated features in `A.B` are removed in `A.B+2` releases (instead of `A+1.0`).

## Cool features are "documented" in the release notes

Meanwhile, we have the "Extending" section of the documentation, which covers customisations to the admin interface. By nature, the features documented here are more "volatile" as the admin is the part we actively work on in every release. We still try to go through the deprecation process where possible, but sometimes we make exceptions when the upgrade path is straightforward enough and keeping the old code is deemed not worth it.

For example, the changes to report views in 6.2 brought the code much closer to how we do listings now (with new designs, AJAX-based filtering, tables framework support). The migration path was relatively straightforward and would bring a lot of benefits to developers, but keeping the old code would be a considerable baggage. The report views were the only place that still used the old non-AJAX based filtering on the right side of the listing, as well as the old listing layout without the slim header – which also had its own issues like #8929, #11198. We'd like to make further improvements to the new designs, but it would be very hard to do if we still had to maintain the old styles. Thus, we made the decision to not go through a deprecation process (https://github.com/wagtail/wagtail/pull/11987#pullrequestreview-2142877059).

On the other hand, sometimes we also take the time to call out breaking changes to undocumented internals in the release notes. As they are completely undocumented, there's no hard rule to follow. We usually make decisions based on how likely the code is used by developers, and whether making a deprecation is worth it instead of removing the old code outright. A common example is our recent deprecations and removals of client-side code that does not comply with Content Security Policy (CSP) on browsers.

In either case, whether it's documented in the "Extending" section or undocumented at all, we often end up documenting both the old and new approaches in the release notes. As a result, when you search for some specific Wagtail admin code in the docs, you may end up finding it in the release notes.

It's [funny (in a cosmic sort of way)](https://youtu.be/DKQ9CrPteDo?t=3) to say that Wagtail's most powerful admin features are documented in the release notes.

## Case for "experimental" features?

Other projects tend to tackle this issue by hiding the features behind "experimental" flags. While this may work well for bigger, separate features that we do not want to enable by default, I don't think it's the right approach for Wagtail.

Our case tends to be changes that we want to take effect immediately for most users. For example: implementing new designs, adding new useful features like the accessibility checker and contextual alt text, and reducing incompatibilities with CSP. While some of these changes involve breaking changes to internal undocumented APIs; for users that don't do any customisation, things remain working as-is and they get the new stuff for free. They are not experimental in the traditional sense.

The "experimental" part for us is the internals that more adventurous users can make use of to extend Wagtail's admin interface. Using experimental flags would mean that basically any new or backwards-incompatible improvement we make to the internals needs to hide behind a flag. Each flag comes with a maintenance burden of its own. The flags would need to be documented too, otherwise you can't expect people to use them.

At that point, the only difference between experimental features and non-experimental features are just how volatile they are, whether they are subject to the deprecation policy, and that they hide behind feature flags. If we want to do that, do we even need the flags?

## Clarifying the deprecation policy

I think the concept of documenting some parts of the codebase that are _not_ subject to the deprecation policy is still useful. To achieve that, we don't need to adopt feature flags – what we need is what they're most useful for: **indicating** that they are subject to change. A "here be dragons" sign, basically.

So I'm proposing that we should update our deprecation policy, and clarify which documented parts are not subject to it. I was thinking of a new section for these features, but given past examples I think this is basically what the "Extending" section is for. Once we make it clear, we can start documenting our internals more freely.

It has benefits for both sides:
- Advanced users can utilise the codebase more easily (with a bit more demanding yet clearer commitment). This can manifest in cool custom features that are published as external packages, or even integrated into core.
- The knowledge of our internals are shared to the world instead of living in the heads of our product team, which makes the project more sustainable. It's also much easier to gather feedback and improve the codebase when you can point people to specific parts of the documentation.

Note that by not including them in our deprecation policy, it doesn't mean that these documented features will break in every release. Given enough time, they may not see any changes over the releases, and are as "stable" as the other parts of the codebase. But, when we _do_ need to make changes, we can be more prepared and clear on what the implications are, as we can refer to the existing documentation.

## Sounds familiar...?

"Are we just going back to the days before 3.0 where we would make breaking changes and remove stuff in non-major releases?"

Well, yes, but also no. For core parts that are documented in the sense of "using" Wagtail, our deprecation policy still takes effect. We're only making it clear that it doesn't apply to the "Extending" Wagtail part of the docs – and giving you more of the docs you need if you want to do it.

## Moving forward

That said, I think the key actions here are:

- [ ] Update our deprecation policy to be more reflective of our practice since v3.0 (#12543)
- [ ] Indicate that features under the "Extending" section of the docs are not subject to the deprecation policy
- [ ] Great to have: #12464 so we can update docs faster
- [ ] Start documenting!

A list of things I have in mind that we can start documenting:

- Generic views, aka "How we build admin views in a post-6.0 world"
  - Update https://docs.wagtail.org/en/stable/extending/generic_views.html to reflect the current state of the art.
- Slim header, how to use them for generic views
  - Breadcrumbs (`breadcrumbs_items`, `get_breadcrumbs_items`)
  - Header buttons (`get_header_buttons`, `get_header_more_buttons`)
  - Side panels! #12179
- Universal listings
  - Tables UI framework
  - Listing buttons
  - Search and filters
  - "Results-only" view
- UI components
  - Dropdowns, buttons, dialogs, etc.
- ViewSet and ViewSetGroup
  - The architecture, and how to pass attributes to views (e.g. `get_FOO_view_kwargs`, `inject_view_methods`). Not just the reference docs, which we already have (and can still improve).
- Choosers?
- Dismissibles
  - Initial "docs" available at https://github.com/wagtail/wagtail/discussions/9351
- Permission policies
- Telepath usage
- Stimulus controllers API reference
- Our future plans and vision for any of the above

There's a lot more that hasn't come to mind. Maybe we could use tools like https://github.com/HunterMcGushion/docstr_coverage?

What documentation would you like to see? Comment below! 😄

**Comment by SebCorbin** [comment]
Well, we just did a presentation at [PyconFr](https://www.pycon.fr/2024/fr/talks/long-talk.html#talk-UYVYQG) (in French) on how we use Wagtail as a framework, so we often use "internals" and when we do: documentation is the code itself :) Hence docstring is indeed useful to us.

And because I'm lazy, let me copy-paste your list and check what we used:

- [x] Generic views, aka "How we build admin views in a post-6.0 world"
- [x] Slim header, how to use them for generic views
  - [x] Breadcrumbs (`breadcrumbs_items`, `get_breadcrumbs_items`)
  - [x] Header buttons (`get_header_buttons`, `get_header_more_buttons`)
  - [ ] Side panels! [Ability to add/modify side panels for pages #12179](https://github.com/wagtail/wagtail/issues/12179) _I need time to work on this, but we just monkey-patch for now_
- [x] Universal listings
  - [x] Tables UI framework
  - [x] Listing buttons
  - [x] Search and filters
  - [x] "Results-only" view
- [x] UI components
  - [ ] Dropdowns
  - [x] Buttons
  - [x] Dialogs
  - [ ] etc.
- [ ] ViewSet and ViewSetGroup
  - [ ] The architecture, and how to pass attributes to views (e.g. `get_FOO_view_kwargs`, `inject_view_methods`). Not just the reference docs, which we already have (and can still improve).
  - _I'm not sure what you mean by that, but most of our viewsets have `{add,edit,inspect}_view_class` attributes overriden_
- [x] Choosers
   - `chosen_view_class`, `choose_view_class`, `choose_results_view_class`, `chosen_multiple_view_class`
- [ ] Dismissibles (not yet :grinning:)
- [x] Permission policies
  -  `ViewOnlyPermissionPolicy`, `ViewEditOnlyPermissionPolicy`, `ViewDeleteOnlyPermissionPolicy`
- [x] Telepath usage
  - this one was kinda hard to put it together, as [current doc](https://docs.wagtail.org/en/stable/reference/streamfield/widget_api.html) lacks an example
- [x] Stimulus controllers API reference
  - as said in a Slack message, we actively use stimulus controllers in our project, namely `w-reveal`  `w-clipboard` `w-progress` `w-action` `w-tooltip` `w-orderable` `w-dialog` `w-slug` (yes not the original one)
  - the TypeScript documentation is quite good, but you have to know where to look

**Reply by lb-** [reply]
As per Slack, it's awesome (or scary) to hear that you're using many of these Stimulus controllers. 

I'm glad the TypeScript / JSDoc is helpful.

**Comment by lb-** [comment]
Adding some thoughts from a core contributor and reviewing perspective. Thanks for raising @laymonage 

## General thoughts

I find it's often harder to get reviews when something is considered public (documented) when either new or changing. There's much more care, intentionally and warranted, to every variable name, docstring and additional edge cases. 

When somethings unofficial, internal, we do try to have some through thinking to these but it's not to the same level.

Sometimes it can mean the difference between weeks or months to get a PR through.

A potential concern of more features being documented by default is it's a higher overhead for all changes to get through. This can also be a high bar for new contributors, or responsibility for reviewing.

However, after introducing my fair share of 'changes to undocumented internals' in the last two years and seeing the bugs raised for 'hey, we were using this and you broke it'. It seems that things will be used anyway, maybe that's an argument for making it easier to use - even if unofficial. 

I don't know the right way forward, having two tiers of what's documented as public/official may be tricky but kind makes sense. However, I think being conservative in what's documented may give us more freedom to evolve the codebase into what organically (ish) gets used more and becomes a candidate for a public API. 

We also want to try to encourage our community to educate us on what they use unofficially, even if that means raising a bug when we change it. This way we know what's used and what needs to get attention.

## Code documentation approach

Anyway, if we were to pursue this, I think we could start with some kind of linting checks for Docstrings / JSDoc. While forcing docs for no reason is unhelpful, I'm sure we can set a baseline of all classes having a description and maybe a status of it's public access. E.g. JSDoc has [`@public`](https://jsdoc.app/tags-public) / [`@private`](https://jsdoc.app/tags-private) / [`@access`](https://jsdoc.app/tags-access).

## Extraction of libraries approach 

Another suggestion, in a different direction, is more intentional splitting of libraries/packages. I'm personally not a big fan of this, I think it makes it harder to contribute and keep things up to date. But for some use cases, it's great, such as the [laces](https://github.com/tbrlpld/laces) package or [`django-modelcluster`](https://github.com/wagtail/django-modelcluster). This approach seems best suited to clearly isolated functionality, with a very stable and small API surface area, the anti-example being [Willow](https://github.com/wagtail/Willow) though.

## Existing requests for 'make this feature documented'

I'm sure this is not an exhaustive list but it's good to flag that we have a range of requests for documenting items. These are often blocked by design decisions needed (usually because we want to be careful about public facing APIs). Maybe getting some headway on known requesta could be good, even a candidate for GSoC. 

Another approach could be to try, in our next survey, get feedback on the most wanted documented features. 

- https://github.com/wagtail/wagtail/issues/11314 (related to ongoing discussion for CSP / Stimulus / Telepath / jQuery plans)
- https://github.com/wagtail/wagtail/issues/8408 (hopefully closer when we have Tabs migrated to Stimulus, I'm getting closer with this)
- https://github.com/wagtail/wagtail/issues/12422
- https://github.com/wagtail/wagtail/issues/12428 (awaiting further input)
- https://github.com/wagtail/wagtail/issues/12179 (might be nice to also have https://github.com/wagtail/wagtail/issues/10168 )
- https://github.com/wagtail/wagtail/issues/11916 (see discussion on issue)
- https://github.com/wagtail/wagtail/issues/11398

**Comment by jhonatan-lopes** [comment]
Thanks @laymonage! This is something that would definitely be helpful!

Adding to the list, I would like to have documentation on how we can extend `ModelViewSet`/`SnippetViewSet` to only support a subset of views, e.g. registering a snippet to only show list and delete views but not the edit view. This is something that comes up frequently for me.

**Comment by thibaudcolas** [comment]
At the size Wagtail is at today (750k LOC, 20 headline features released per year, hundreds of packages), I think we also need some of those docs to simply be able to stay ahead of technical debt, and reduce the project’s bus factor (lots of code that few people are familiar with, that takes a while to understand the current state of, and desired state)

For me a good example is UI components, seeing that we’ve invested a lot in refactoring them to be more reusable, internally and for packages. Obvious ones:

1. Buttons
2. Tooltips
3. Dropdowns
4. Dialogs
5. Status tags

Less obvious but still important

- Collapsible panels
- Tabs
- Progress indicators
- Keyboard shortcuts
- Avatar

Related docs:

- Browser support updates (https://github.com/wagtail/wagtail/pull/13197#discussion_r2182169198)

For all of those I think there’s a case to document how to use them both as far as the intended UX ("always add a tooltip to an icon-only button") as well as the coding.

---

I think we should also consider this documentation work through the lens of AI-assisted development. AI in pull requests creation, AI in PR reviews, AI in QA. We don’t want people to think they can vibe code an open source contribution but there’s clearly potential in AI becoming more aware of a project’s specific contribution guidelines to create more correct implementations.

**Comment by thibaudcolas** [comment]
Oh one last thought – I think this type of thread is a good indicator that we should curate a longer-lasting backlog of tech debt / project health improvements. Most likely in GitHub issues? We’ve been keeping tabs on issue counts as a metric of project health in the past, but really it’d be worthwhile to separately consider bug reports, feature requests, and refactorings / maintenance tasks.

**Comment by thibaudcolas** [comment]
Moving forward with UI component documentations 👉 #13207. With work from @laymonage to integrate with typedoc specifically: #13213. Direct link to the [sphinx.js + typedoc integration](https://wagtail--13213.org.readthedocs.build/en/13213/reference/ui/client/index.html).

Feedback very welcome! For now I think our plan is to proceed with something very extensive and automated like that and see what happens. But we will still need to figure out how best to convey which components are stable and intended for reuse, and which ones aren’t stable or are meant to be private.

**Comment by laymonage** [comment]
Update on this: documentation for the Tables framework was added in Wagtail 7.4 🎉 

https://docs.wagtail.org/en/stable/reference/ui/tables.html
